richer api errors proposal

ripienaar opened the pull request:
This creates richer errors in response to https://github.com/nats-io/nats-server/issues/1811 and numerous other requests.

The main idea is that when a 404 is raised in response to accessing a consumer we now include a unique identifier that would allow callers to distinguish between stream not found, consumer not found, no message found and other errors that might raise 404.

One could argue that users might parse the Description like `stream not  found` however this makes our human friendly descriptions part of the API promises and we can never improve, bug fix or translate those strings. For example today there is an unfortunate mix of `jetstream` and `JetStream` in use in error descriptions and we would want the ability to improve errors without breaking peoples code.

Additionally I propose capturing help and urls in the code - not in actual API responses - to facilitate a single place where we maintain these errors and then adding tooling like `nats error JS1` that will produce the error text, help, additional information and links to documentation that would help users resolve those errors.  So the slow `ApiErrorByKind` would not be used in anything but the CLI to resolve these things, though even thats optional as the `ApiErrors` is exported.

Signed-off-by: R.I.Pienaar <rip@devco.net>

/cc @nats-io/core

ripienaar commented on server/jetstream_api.go:
```diff
@@ -608,6 +636,17 @@ var (
+func ApiErrorByKind(kind string) *ApiError {
+	for _, v := range ApiErrors {
+		if v.Kind == kind {
+			return v
```
They are unique yes. 

My thinking is that when writing or reading:

```
if !acc.JetStreamEnabled() {
		resp.Error = jsNotEnabledErr
		resp.Error = ApiErrors[JSNotEnabledErr]
		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
		return
}
```

is easier to parse while scanning code due to the `JSNotEnabledErr` token than:

```
if !acc.JetStreamEnabled() {
		resp.Error = jsNotEnabledErr
		resp.Error = ApiErrors[JS1]
		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
		return
}
```

This is perhaps not the best example but we have cases where we have like 3 or more if statements quite densely returning different errors.  For me when scanning such code I look for log lines and error strings rather than code - this retains a better readability to my eye

ripienaar replied:
Also you get the benefit of IDE completing errors for you and you dont need to remember whats the code for an error, you see error constants.

ripienaar replied:
Here is an earlier stab I took at doing this - we have many many errors:

https://github.com/ripienaar/nats-server/blob/cdef3e8acb07565336efda3c89b2e51be696bb99/server/jetstream_api.go#L600

I dont want to remember Kind codes.

kozlovic replied:
I did not make myself clear, I was saying that we can have 2 maps, one keyed by the id like you have (say JSNotEnabledErr, etc), and another that will be keyed by the kind ("JS01", "JS02", etc..). As you can see from my code snippet, I use both maps.
The advantage of the 2nd map (keyed by kind) is that there is no longer "slow lookup". The range and checking for kind basically becomes:
```
func ApiErrorByKind(kind string) *ApiError {
    return ApiErrorsByKind[kind]
}
```

ripienaar replied:
Sure, I dont think we have a need in the server for looking up by kind really and I dont want to forget to maintain one map all the time.  Can you think of a reason in the server to find by kind?

The CLI will loop when needed but its so fast - few 100 keys max - its effectively instantaneous.  I'd just remove the lookupbykind - there's no use. 

CLI will need to loop anyway as I want to be able to search/list/etc the kinds and so forth

kozlovic replied:
Oh ok, I got confused, I thought that you would get this lookup in the server that the server could query, etc..

ripienaar replied:
Yeah, no I dont think we really need that in the server

ripienaar commented:
@derekcollison I want to spend some time on this during this week, let me know what you think of the approach before I dive in on a wrong track, thanks!

derekcollison commented:
Probably deserves a meeting between a few folks.

ripienaar commented:
I added @matthiashanel and @aricart to reviewers for context, will have a call with all reviewers I guess.

derekcollison commented:
Add me onto the list as well if you can.

aricart commented:
I had looked at the PR before, and I agree that having a code that can be pointed at documentation can provide a richer experience to diagnose problems. The standard string that it represents is useful, but more so I think is the case where these values make it to a server log. Having documentation where the lookup can be performed is very useful.

derekcollison commented on server/jetstream_api.go:
```diff
@@ -259,7 +259,22 @@ const JSMaxNameLen = 256
 // TODO(dlc) - Move to more generic location.
 type ApiError struct {
 	Code        int    `json:"code"`
+	Kind        string `json:"kind,omitempty"`
```
Pick different name?

Internal
ErrCode
Err

derekcollison commented on server/jetstream_api.go:
```diff
@@ -259,7 +259,22 @@ const JSMaxNameLen = 256
+	return &ApiError{
+		Code:        e.Code,
+		Kind:        e.Kind,
+		Description: fmt.Sprintf(e.Description, a...),
```
We can do the camel case here. Most are fine but if we wrap an err then staticcheck forces those to be all lowercase.

derekcollison commented on server/jetstream_api.go:
```diff
@@ -588,8 +603,21 @@ type JSApiStreamTemplateNamesResponse struct {
+	ApiErrors = map[APIErrorIdentifier]*ApiError{
+		JSUnknownErr:                    {Code: 503, Description: "Unknown error", Kind: "JS0"},
+		JSNotEnabledErr:                 {Code: 503, Description: "JetStream not enabled for account", Kind: "JS1", Help: "This error indicates that JetStream is not enabled either at a global level or at global and account level", URL: "https://docs.nats.io/jetstream"},
+		JSExternalDelPrefixOverlapsErrF: {Code: 400, Description: "stream external delivery prefix %q overlaps with stream subject %q", Kind: "JS2"},
```
Cap stream ?

ripienaar commented on server/jetstream_api.go:
```diff
@@ -499,36 +483,31 @@ const JSApiMsgGetResponseType = "io.nats.jetstream.api.v1.stream_msg_get_respons
 
 const JSApiConsumerDeleteResponseType = "io.nats.jetstream.api.v1.consumer_delete_response"
 
-// JSApiConsumerInfoResponse.
```
these are more or less useless comments so just remove them, they dont even make my IDE not complain

derekcollison replied:
What about `go vet` and `staticcheck`?

kozlovic replied:
Used that linters on my IDE would complain, but not any more. go vet and staticcheck don't seem to have a problem with those either. So agreed that if we are not adding more meaningful comment, then better removed.

ripienaar commented on server/jetstream_errors.go:
```diff
@@ -0,0 +1,220 @@
+
+const (
+	JSUnknownErr APIErrorIdentifier = iota
+	JSAccountResourcesExceededErr
```
I kind of want to generate this file rather than hand maintain it. Given a source - json or yaml - that roughly looks like the structure a bit lower `go generate` would make this file.

The benefit of this is that I could automatically add a comment here above each of these identifiers that include the Description from the `ApiErrors` struct, this will greatly help code use because when you hover your mouse over the error you'll get some feedback about its body and such, now using these are a bit of pain.  

I doubt yous will like `go generate` based flow though.

derekcollison replied:
I am not a fan of it as a main build flow, but we could use it and check in the resulting files.

ripienaar replied:
yes, that would be the idea - use it only when the source file changes and check in the result

ripienaar replied:
Discussed with Ivan, we'll go with generate

ripienaar commented on server/jetstream_errors.go:
```diff
@@ -0,0 +1,220 @@
+
+var (
+	ApiErrors = map[APIErrorIdentifier]*ApiError{
+		JSUnknownErr:                       {Code: 503, ErrCode: "JS001", Description: "Unknown error"},
```
Not decided yet if `JS001` is the right format, feedback welcome, I do think strings not numbers are better and the catagorisation that the `JS` adds is nice.

derekcollison replied:
Numbers are directly comparable in all languages, that could be an advantage. Interested in what other's think.

kozlovic replied:
Numbers would be easier indeed. But if the category is useful (JS), it could be made an ErrCategory: "JS" for people who care to know. But the ErrCode alone (a number) would be enough to check for the error code. However, for applications to be able to ignore the category means that error codes should be ever increasing, that is, do not repeat err codes because they belong to different categories.

PS: I get that Code already exists in ApiError, and RI proposed Kind first. But ErrCode is a bit of a stutter, as in ApiError.ErrCode.

ripienaar replied:
See comments below about how to check for errors etc, but I agree probably `JS001` is not needed.  How would you feel about codes like `1xxxx` being JS and `2xxxx` being MQTT etc?

ripienaar commented on server/jetstream_errors.go:
```diff
@@ -0,0 +1,220 @@
+	}
+}
+
+func (e *ApiError) ErrOrNewF(err error, a ...interface{}) *ApiError {
```
will add some comments

derekcollison replied:
I am the worst at naming but might consider something a bit more descriptive.

ripienaar replied:
I'd need some suggestions, as these are already attempts at improving the situation :)

derekcollison replied:
Describe the function for us.

ripienaar replied:
Uncommited locally I have this comment

```
// ErrOrNewF returns err if it's an ApiError else creates a new error using Sprintf
```

ripienaar replied:
```
err = acc.deleteStreamTemplate(name)
if err != nil {
		resp.Error = ApiErrors[JSStreamTemplateDeleteErrF].ErrOrNew(err)
		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
		return
}
```

Here if deleteStreamTemplate() returns an apiError - that one will be used, otherwise a new `JSStreamTemplateDeleteErrF` is made via Sprintf

This way we can return a ApiError for things like template not found but a io error if deleting it from disk fails, and the fact that the problem is template not found would be surfaced correctly.

ripienaar commented on server/jetstream_test.go:
```diff
@@ -7417,17 +7417,15 @@ func TestJetStreamRequestAPI(t *testing.T) {
 	}
 
-	checkBadRequest := func(e *ApiError, description string) {
+	checkNatsError := func(e *ApiError, id APIErrorIdentifier) {
```
do we have a more generic place for test helper functions? I suspec this will be used quite a lot

kozlovic commented on server/mqtt.go:
```diff
@@ -959,7 +959,7 @@ func (s *Server) mqttCreateAccountSessionManager(acc *Account, quitCh chan struc
 		Replicas:  as.replicas,
 	}
-	if _, err := jsa.createStream(cfg); isErrorOtherThan(err, ErrJetStreamStreamAlreadyUsed) {
+	if _, err := jsa.createStream(cfg); isErrorOtherThan(err, ApiErrors[JSStreamNameExistErr]) {
```
Had this been in place, I would have checked for an error code here instead. The helper(s) createStream() and the like could be changed to return an error based on ApiError, but also the error code and that is what I would check here: (errors should be returned as last argument)
```
if _, errCode, err := jsa.createStream(cfg); errCode != "JS058" {
    return nil, fmt.Errorf("...", acc.GetName(), err)
}
```
(above is based on currently errCode be the ApiError.ErrCode which is a string as opposed to a number)

ripienaar replied:
It seems quite a verbose pattern to also return a code there - making go's already awful error handling more verbose and more returned values etc. 

Mainly I want to make as little as possible burdon on those writing functions, shifting the effort instead on their callers - most callers would never need to check a specific err code.

```
if _, err := foo(); !IsNatsErr(err, "JS058") { ... }
```

Wouldnt overly complicate the functions, allow us to use richer errors progressively in functions (what if `foo()` returned a ApiErr sometimes othertimes a go error? which is happening a bit in functions already).  Probably about the same level of effort but feels more go like.

ripienaar commented:
@kozlovic think I am good on this one, I updated MQTT code considering new errors and also added a ADR for how it works and how to maintain it (though I think that needs expanding in time).

for now keen to get this in so we can stop the rebasing pain :)

kozlovic commented on server/errors.go:
```diff
@@ -171,46 +171,10 @@ var (
 	ErrNoTransforms = errors.New("no matching transforms available")
-
-	// ErrJetStreamNotEnabled is returned when JetStream is not enabled.
-	ErrJetStreamNotEnabled = errors.New("jetstream not enabled")
```
As we discussed once, should we leave those but with a doc section that says that they are "deprecated"? I am thinking of use-case where users embed the NATS Server. I assume that internally they will no longer be used or updated.

ripienaar replied:
Yeah its a difficult one, we could say `ErrJetStreamNotEnabled = ApiErrors[JSWhatever]` but I am not 100% sure that will solve the problem in all cases tbh

ripienaar replied:
Actually it should solve it for these - as they are essentially static, the ones where it wont solve it are the ones where we put in a different description via printf.  So it might be fine

derekcollison replied:
We should leave, no harm imo..

kozlovic replied:
I think the point was that we would not use them anymore, but we may want to keep them as to possibly not break users that embed NATS Server and somehow were checking those error codes?

ripienaar replied:
Yeah, I ppreciate why we would want to keep them - however, keeping them would not solve the problem completely.

For some we can keep them via an alias asignment but any of them that were dynamic would not be possible (the ones with tokens in them):

```
ErrJetStreamNotEnabled = ApiErrors[Foo]
```

Non of these had tokens of course so I need to look, might have to adjust `NewT()` and related functions to instead of a new instance return the existing pointer that is in `ApiErrors`, thats the only way it could possibly work.  This though is risky someone could adjust the "source" of these errors and bad things will happen, as there's no way to really make them read only as such

kozlovic replied:
Yes, it silly of me to say to keep them not to "break" their code (because say `ErrJetStreamNotEnabled` is no longer defined), while it would break it in worse way by having the check for say `ErrJetStreamNotEnabled` be false because of the new errors returned.
So I would be ok if we remove in 2.3.0, but then if we have to release a 2.2.7 before that release, then we would have to create and maintain a `branch_2_2_x` to release from, which I would rather not have to do..

ripienaar replied:
Old costants were restored in https://github.com/nats-io/nats-server/pull/2168/commits/ee9d10f40b8f16aacddd57f8142df7004422a532 with appropriate comments warning people etc

I think this one is good to go now

kozlovic reviewed (approved):
LGTM, except not sure what to do with exported JS errors: removing them may break some apps (but unlikely).

ripienaar commented:
While we wait for @derekcollison to give this a look the one thing I've been mulling over is how we format the errors that are of the ErrorF form.

```
ApiErrors[JSStreamExternalApiOverlapErrF].ErrOrNewF(foo, bar)
```

Where the ApiErr is:

```
{Code: 400, ErrCode: 10021, Description: "stream external api prefix %q must not overlap with %s"},
```

The  problem here is that we want to be able to extend and change these errors - especially if we later need to translate them - then the order based replace becomes an error.

The typical fix is to make the description something like `"stream external api prefix {prefix} must not overlap with {api}"` and then:

```
ApiErrors[JSStreamExternalApiOverlapErrF].ErrOrNewF("{prefix}", foo, "{api}", bar)
```

This is much more robust and future proof, personally I think it's worth it, but curious what @ColinSullivan1 (who like me also seem to care for translation) and @kozlovic thinks

kozlovic commented:
> This is much more robust and future proof, personally I think it's worth it, but curious what @ColinSullivan1 (who like me also seem to care for translation) and @kozlovic thinks

I agree.

derekcollison commented on server/jetstream_api.go:
```diff
@@ -262,7 +262,22 @@ const JSMaxNameLen = 256
 	Code        int    `json:"code"`
+	Kind        string `json:"kind,omitempty"`
 	Description string `json:"description,omitempty"`
+	URL         string `json:"-"`
```
Why are these never serialized and sent to clients?

ripienaar replied:
These are there mainly to assist tooling - I will add some `nats explain` or something that read these and generate documents, links to further information etc.

Doesnt seem appropraite to put in API responses - its a documentation link.

derekcollison replied:
Should these be separate aggregate struct then that extends?

ripienaar replied:
Sure, but whats the benefit? I don't see how this is better or worse then?

derekcollison replied:
Feels odd to have public members on a struct used to send JSON responses that includes things that we do not send.. But not super important. I would do aggregate (has-a) or not make them public, but again not that critical.

ripienaar replied:
They are public so that other tools - like the NATS CLI can get access to them via the Go API rather than have to maintain some additional structures etc.

So for that use case I think this is exactly where they should be, even if I added them in another struct that extend, to allow the CLI to access that extending struct I would have to export that struct key and tell JSON to ignore it. Which feels about exactly the same as this.

derekcollison replied:
Not following that part, so you are using the NATS server source code directly in the NATS cli or do you mean the Go client? Which is of course very different and will have its own version of this.

ripienaar replied:
Only the CLI

Definitely not any clients - they only need to know the ErrCode which is in the JSON.

(something like)
```
$ nats explain 10039
Message:  JetStream not enabled for account
Help: https://docs.nats.io/jetstream

This error indicates that JetStream is not enabled either at a global level or at global and account level
```

This help url and additional short help comes out of these structs directly

ripienaar replied:
However, to be exactly clear about why I did certain things, these errors are all in a JSON files in this repository. The build process of the CLI can pull in that json file and generate the richer structures there so this is all out of the nats server go code but we still end up with one place to maintain the data.

So I can completely remove these fields from the structs here and it would be ok(ish) though more work in the CLI.  But the JSON file is there so that any other language that wish to gain access to this stuff can do so

derekcollison replied:
How do we plan on dealing with server / cli version mismatch in this case? Should the json be something you can get from the server's system account maybe?

ripienaar replied:
We could for sure do that, we do have quite a bit of version coupling between CLI and Server atm so don't consider it the end of the world.

It's probably something we can improve around the time we do schema registeries - so this is some data that can also provide?

derekcollison replied:
Might be worth discussing.

derekcollison commented on server/jetstream_api.go:
```diff
@@ -262,7 +262,22 @@ const JSMaxNameLen = 256
+}
+
+func (e *ApiError) Error() string {
+	return e.Description + " (" + e.Kind + ")"
```
Not sure if we expect for this to be called alot but this generates a bunch of temp objects etc. Might be better to do StringBuilder.

ripienaar replied:
not a lot but not that its a uint there we also have to do strconv etc - so I think just using fmt.Sprintf() is going to be best.

derekcollison commented on server/jetstream_api.go:
```diff
@@ -262,7 +262,22 @@ const JSMaxNameLen = 256
 // TODO(dlc) - Move to more generic location.
 type ApiError struct {
 	Code        int    `json:"code"`
+	Kind        string `json:"kind,omitempty"`
```
We should think if these should be strings or numbers. For direct comparison we would have to look to C and would need numbers, not strings.

derekcollison commented on server/jetstream_errors.go:
```diff
@@ -84,100 +84,115 @@ const (
 type ApiError struct {
 	Code        int    `json:"code"`
-	Kind        string `json:"kind,omitempty"`
+	ErrCode     string `json:"err_code,omitempty"`
```
I think this should be u16.

ripienaar replied:
agree,

Not sure why I cant comment directly by the:

> We should leave, no harm imo..

I think there is plenty of harm tbh, in order to return the code we have to use a different type - we wont use this constant anymore so it'll just cause confusion.

We could as I say do `ErrJetStreamNotEnabled = ApiErrors[xxx]` but that only works for the kind of error without any dynamic content so would work in this specific case but in general not.

derekcollison replied:
ok maybe you are correct and since we made the functions private no one could really run something and use those.

derekcollison reviewed (approved):
In general LGTM. Went through them one by one so ignore repeated comments on codes need to be numbers.

ripienaar commented:
I squashed this but I think I need more time to find a better solution to handle the named token replace things, forcing all things to be strings is terrible which is what strings.Replacer() wants.  So will investigate more.

derekcollison commented:
We could do the string replace solution at a different time as well.

ripienaar commented on server/jetstream_errors.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+// NewT creates a new error using strings.Replacer on the Description field, arguments must be an even number like NewT("{err}", err)
+func (e *ApiError) NewT(replacements ...interface{}) *ApiError {
```
@kozlovic So I added tokenization in 40c9c1c and updated all the things and rebased to master.  wdyt?

derekcollison reviewed (commented):
I think we can merge once @kozlovic has signed off.

kozlovic commented on doc/adr/0007-error-codes.md:
```diff
@@ -0,0 +1,142 @@
+|Range|Description|
+|-----|-----------|
+|1xxxx|JetStream related errors|
+|2xxxx|MQTT related errors|
```
To be clear, there is no such error at the moment...

kozlovic reviewed (commented):
Last comment regarding removing server's existing `ErrXXX`...

kozlovic commented on server/errors.json:
```diff
@@ -187,9 +188,17 @@
   {
     "constant": "JSNotEnabledErr",
     "code": 503,
+    "error_code": 10076,
```
Changing the error code after this PR is merged is something that we can't be doing, correct?

This file is manually maintained, correct? I see that the error codes are a bit all over the places, which may be error prone when adding a new error to pick a code that is already used. But I guess the generator would check on that and fail if that is the case?

ripienaar replied:
Yes, can not change.  The file is indeed a pain do you sort it by code, constant or what? :)

Generater does check and fail, I might make it report whats the highest current error code in that check though

ripienaar replied:
Updated thanks, I imagine when we add other subsystems I'll need a tool to maintain this file

kozlovic commented on server/errors.json:
```diff
@@ -187,9 +188,17 @@
+    "error_code": 10076,
+    "description": "JetStream not enabled",
+    "help": "This error indicates that JetStream is not enabled at a global level",
+    "deprectes": "ErrJetStreamNotEnabled"
```
Typo? `deprectes`

kozlovic reviewed (changes requested):
New field `deprecates` is misspelled in one of the error, also comment regarding ordering of the error codes in the json file.

kozlovic commented on server/errors_gen.go:
```diff
@@ -74,16 +74,20 @@ func goFmt(file string) error {
 	for _, entry := range errs {
 		if _, found := codeKeys[entry.ErrCode]; found {
-			return fmt.Errorf("duplicate error code %+v", entry)
+			return fmt.Errorf("duplicate error code %+v, highest code is %d", entry, highest)
```
Good, but I thought that you would have an option to the tool to tell us what are the highest codes for each "category". Can be added later, so not holding the merge for that.

ripienaar replied:
Indeed, once we have more than one cat I will add :P
